pkg/controller/templateupdaterequest/templateupdaterequest_controller.go

```diff
@@ -0,0 +1,320 @@
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, _ *configuration.Config) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileTemplateUpdateRequest{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("templateupdaterequest-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource TemplateUpdateRequest
+	err = c.Watch(&source.Kind{Type: &toolchainv1alpha1.TemplateUpdateRequest{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileTemplateUpdateRequest implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileTemplateUpdateRequest{}
+
+// ReconcileTemplateUpdateRequest reconciles a TemplateUpdateRequest object
+type ReconcileTemplateUpdateRequest struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a TemplateUpdateRequest object and makes changes based on the state read
+// and what is in the TemplateUpdateRequest.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileTemplateUpdateRequest) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	logger.Info("Reconciling TemplateUpdateRequest")
+
+	// Fetch the TemplateUpdateRequest tur
+	tur := &toolchainv1alpha1.TemplateUpdateRequest{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, tur)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// if is complete, then check when status was changed and delete it if the requested duration has passed
+	completeCond, found := condition.FindConditionByType(tur.Status.Conditions, toolchainv1alpha1.TemplateUpdateRequestComplete)
+	if found && completeCond.Status == corev1.ConditionTrue {
```

Comment:
so what's the point of logging the error in the status in the CR, then?

Reply:
it's the information for `NSTemplateTier` so it is able to track the failed updates. And as I have written above, it should be also a sign that the `TemplateUpdateRequest` is ready to be deleted (as soon as the `NSTemplateTier` controller adds it to the list of the failed updates)

Reply:
but still, as I have written - we have to remove them at some point - the `TemplateUpdateRequest` CRs have the names of the MUR, so if we kept the failed `TemplateUpdateRequest` CRs in the cluster, then we wouldn't be able to update the MURs next time

Reply:
yes, I know we can't run another update if we have remaining `TemplateUpdateRequest` resources in the namespace.

But if we have dozens of failures (let's say, for a reason we did not foresee), are we going to store them all in the `NSTemplateTier`?!? Isn't it there a risk of making the resource huge? Also, what's the policy for cleaning these "logs" afterwards? 

IMO, keeping the `TemplateUpdateRequest` in failed state and having an admin take care of analyzing the errors and deleting the resources afterwards is a valid solution, too.

Reply:
> But if we have dozens of failures (let's say, for a reason we did not foresee), are we going to store them all in the NSTemplateTier?!? Isn't it there a risk of making the resource huge? 

but it's still a list of MUR names, not the whole status or conditions, only the names. That won't make the  `NSTemplateTier` so huge. Anyway, if there are  dozens of failures then the update should automatically stop because there is apparently some significant problem (but that is a feature for later)

>   Also, what's the policy for cleaning these "logs" afterwards?

how I would do it - as inspiration I would use the `ClusterVersion` CRD that is in every cluster - the status would have two parts (or maybe more)
* one part contains information about the current update (or the last one that has finished/failed)
  * this contains all necessary information for an ongoing update
  * contains the number of successful updates 
  * contains a list of MURs for which the update failed 
* second part contains history of the previous updates
  * it contains only simple information about the updates that happened in the past (time, result, number of failed, the hash of the tier,...)

when there is a new update started, then:
1. takes data from the part for the last/current update
2. adds an entry in the history part
3. cleans the part for the last/current update

> IMO, keeping the TemplateUpdateRequest in failed state and having an admin take care of analyzing the errors and deleting the resources afterwards is a valid solution, too.

I'm a bit afraid of the manual deletion:
* we could forget to delete it before the next update
* we could delete `TemplateUpdateRequest` that is related to a different tier
* the deletion would be manual and one-by-one (as you said - there is no way to filter them by a status)
* we could accidentally delete something else :-)

Reply:
I suggest we address all the "monitoring/progress" in a separate story/PR. This one is quite big already and I'd like to see if merged, rather than adding more feature.

Reply:
This will be addressed in https://issues.redhat.com/browse/CRT-711

Reply:
sounds good :+1:

Reply:
thanks ;)